File: conductor/src/main/java/be/vergauwen/simon/conductor/ui/adapter/ItemAdapter.java
package be.vergauwen.simon.conductor.ui.adapter;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import be.vergauwen.simon.common.di.model.Item;
import be.vergauwen.simon.conductor.R;
import kotlin.collections.CollectionsKt;

public class ItemAdapter extends RecyclerView.Adapter<ItemViewHolder> {

    private final List<Item> items = new ArrayList<>();

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public ItemViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        return new ItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_view, parent, false));
    }

    @Override
    public void onBindViewHolder(final ItemViewHolder holder, final int position) {
        holder.bind(items.get(position));
    }

    public void addItem(@NonNull final Item item) {
        if (!items.contains(item)) {
            items.add(item);
            notifyItemInserted(items.indexOf(item));
        }
    }

    public @Nullable Item getItem(final int position) {
        return CollectionsKt.getOrNull(items,position);
    }
}

File: app/src/main/kotlin/be/vergauwen/simon/konductor/ui/adapter/ItemAdapter.kt
package be.vergauwen.simon.konductor.ui.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import be.vergauwen.simon.konductor.R
import be.vergauwen.simon.konductor.core.anko.itemIconView
import be.vergauwen.simon.konductor.core.anko.setMinimumListHeight
import be.vergauwen.simon.konductor.core.anko.setSelectableItemBackground
import be.vergauwen.simon.konductor.core.model.Item
import org.jetbrains.anko.*

class ItemAdapter(private val items: MutableList<Item> = mutableListOf<Item>()) : RecyclerView.Adapter<ItemViewHolder>() {

    ///////////////////
    // Adapter Impl

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder = ItemViewHolder(buildListItemUI(parent.context))

    override fun onBindViewHolder(holder: ItemViewHolder?, position: Int) {
        holder?.bind(items[position])
    }

    override fun getItemCount(): Int = items.size

    ///////////////////

    fun addItem(item: Item) {
        if (items.contains(item)) return
        items.add(item)
        notifyDataSetChanged()
    }

    fun getItem(position: Int): Item = items[position]

    ///////////////////
    // Anko View

    private fun buildListItemUI(context: Context): View =
            context.UI {
                linearLayout {
                    lparams(width = matchParent) {
                        padding = dimen(R.dimen.activity_horizontal_margin)
                    }
                    setSelectableItemBackground()
                    setMinimumListHeight()

                    itemIconView {
                        id = R.id.action_icon
                        gravity = Gravity.CENTER
                    }.lparams(width = 0, height = wrapContent, weight = 1f) {
                        marginEnd = dip(16f)
                    }

                    textView {
                        id = R.id.action_text
                        gravity = Gravity.START
                    }.lparams(width = 0, height = wrapContent, weight = 4f)
                }
            }.view
}